src/parquet/arrow/writer.cc

```diff
@@ -533,65 +545,132 @@ Status FileWriter::Impl::WriteNullableBatch<Int96Type, ::arrow::TimestampType>(
-NULLABLE_BATCH_FAST_PATH(Int32Type, ::arrow::Date32Type, int32_t)
-NULLABLE_BATCH_FAST_PATH(Int64Type, ::arrow::Int64Type, int64_t)
-NULLABLE_BATCH_FAST_PATH(Int64Type, ::arrow::TimestampType, int64_t)
-NULLABLE_BATCH_FAST_PATH(Int64Type, ::arrow::Time64Type, int64_t)
-NULLABLE_BATCH_FAST_PATH(FloatType, ::arrow::FloatType, float)
-NULLABLE_BATCH_FAST_PATH(DoubleType, ::arrow::DoubleType, double)
+Status FileWriter::Impl::WriteTimestamps(ColumnWriter* column_writer,
+                                         const std::shared_ptr<Array>& values,
+                                         int64_t num_levels, const int16_t* def_levels,
+                                         const int16_t* rep_levels) {
+  const auto& type = static_cast<::arrow::TimestampType&>(*values->type());
+
+  const bool is_nanosecond = type.unit() == TimeUnit::NANO;
+
+  if (is_nanosecond && arrow_properties_->support_deprecated_int96_timestamps()) {
+    return TypedWriteBatch<Int96Type, ::arrow::TimestampType>(
+        column_writer, values, num_levels, def_levels, rep_levels);
+  } else if (is_nanosecond ||
+             (arrow_properties_->coerce_timestamps_enabled() &&
+              (type.unit() != arrow_properties_->coerce_timestamps_unit()))) {
+    // Casting is required. This covers several cases
+    // * Nanoseconds -> cast to microseconds
+    // * coerce_timestamps_enabled_, cast all timestamps to requested unit
+    return WriteTimestampsCoerce(column_writer, values, num_levels, def_levels,
+                                 rep_levels);
+  } else {
+    // No casting of timestamps is required, take the fast path
+    return TypedWriteBatch<Int64Type, ::arrow::TimestampType>(
+        column_writer, values, num_levels, def_levels, rep_levels);
+  }
+}
 
-Status FileWriter::Impl::TypedWriteBatchConvertedNanos(
-    ColumnWriter* column_writer, const std::shared_ptr<Array>& array, int64_t num_levels,
-    const int16_t* def_levels, const int16_t* rep_levels) {
+Status FileWriter::Impl::WriteTimestampsCoerce(ColumnWriter* column_writer,
+                                               const std::shared_ptr<Array>& array,
+                                               int64_t num_levels,
+                                               const int16_t* def_levels,
+                                               const int16_t* rep_levels) {
   // Note that we can only use data_buffer_ here as we write timestamps with the fast
   // path.
   RETURN_NOT_OK(data_buffer_.Resize(array->length() * sizeof(int64_t)));
   int64_t* data_buffer_ptr = reinterpret_cast<int64_t*>(data_buffer_.mutable_data());
 
-  auto data = static_cast<const NumericArray<::arrow::TimestampType>*>(array.get());
-  auto data_ptr = reinterpret_cast<const int64_t*>(data->values()->data());
+  const auto& data = static_cast<const ::arrow::TimestampArray&>(*array);
+  auto data_ptr = reinterpret_cast<const int64_t*>(data.values()->data())
```

Comment:
We can do that. I also opened ARROW-1335 because `TimestampArray::raw_values` includes the slice offset while `PrimitiveArray::raw_values` does not (both are used in this file, I think)